pkg/distributor/forwarding/pools.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package forwarding
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// pools is the collection of pools which the forwarding uses when building remote_write requests.
+// Even though protobuf and snappy are both pools of []byte we keep them separate because the slices
+// which they contain are likely to have very different sizes.
+type pools struct {
+	protobuf    sync.Pool
+	snappy      sync.Pool
+	request     sync.Pool
+	bytesReader sync.Pool
+	tsByTargets sync.Pool
+
+	// Mockable for testing.
+	getProtobuf    func() *[]byte
+	putProtobuf    func(*[]byte)
+	getSnappy      func() *[]byte
+	putSnappy      func(*[]byte)
+	getReq         func() *request
+	putReq         func(*request)
+	getBytesReader func() *bytes.Reader
+	putBytesReader func(*bytes.Reader)
+	getTsByTargets func() tsByTargets
+	putTsByTargets func(tsByTargets)
+	getTs          func() *mimirpb.TimeSeries
+	putTs          func(*mimirpb.TimeSeries)
+	getTsSlice     func() []mimirpb.PreallocTimeseries
+	putTsSlice     func([]mimirpb.PreallocTimeseries)
+}
+
+func newPools() *pools {
+	p := &pools{
+		protobuf:    sync.Pool{New: func() interface{} { return &[]byte{} }},
+		snappy:      sync.Pool{New: func() interface{} { return &[]byte{} }},
+		request:     sync.Pool{New: func() interface{} { return &request{} }},
+		bytesReader: sync.Pool{New: func() interface{} { return bytes.NewReader(nil) }},
+		tsByTargets: sync.Pool{New: func() interface{} { return make(tsByTargets) }},
+	}
+
+	p.getProtobuf = getProtobuf(&p.protobuf)
+	p.putProtobuf = putProtobuf(&p.protobuf)
+	p.getSnappy = getSnappy(&p.snappy)
+	p.putSnappy = putSnappy(&p.snappy)
+	p.getReq = getReq(&p.request)
+	p.putReq = putReq(&p.request)
+	p.getBytesReader = getBytesReader(&p.bytesReader)
+	p.putBytesReader = putBytesReader(&p.bytesReader)
+	p.getTsByTargets = getTsByTargets(&p.tsByTargets)
+	p.putTsByTargets = putTsByTargets(&p.tsByTargets)
```

Comment:
Not a blocking nitpick, but we're already compiling with go 1.18, so if we enabled it in `go.mod` this is a good candidate for generics, as we can cover everything with just two functions:

```go
func getter[T any](pool *sync.Pool) func() T {
	return func() T {
		return pool.Get().(T)
	}
}

func putter[T any](pool *sync.Pool) func(T) {
	return func(val T) {
		pool.Put(val)
	}
}
```

And change this to:
```suggestion
	p.getProtobuf = getter[*[]byte](&p.protobuf)
	p.putProtobuf = putter[*[]byte](&p.protobuf)
	p.getSnappy = getter[*[]byte](&p.snappy)
	p.putSnappy = putter[*[]byte](&p.snappy)
	p.getReq = getter[*request](&p.request)
	p.putReq = putter[*request](&p.request)
	p.getBytesReader = getter[*bytes.Reader](&p.bytesReader)
	p.putBytesReader = putter[*bytes.Reader](&p.bytesReader)
	p.getTsByTargets = getter[tsByTargets](&p.tsByTargets)
	p.putTsByTargets = putter[tsByTargets](&p.tsByTargets)
```

Reply:
I love that idea, but I don't want to change the version in `go.mod` within this request. I'll make a separate one and then once it's merged I'll rebase this one.

Reply:
This is so much better ❤️ 

https://github.com/grafana/mimir/blob/81aefe52b7c72976c8d7b7c4f0c56362ad7411d7/pkg/distributor/forwarding/pools.go#L48-L61